### ISSUE TITLE
Core panic with MEMTAG=y

### ISSUE THREAD

**Issue by jforissier** [OP-TEE/optee_os#5759]
On [QEMUv8](https://optee.readthedocs.io/en/latest/building/devices/qemu.html#qemu-v8) with `make MEMTAG=y run`, OP-TEE panics on boot:
```
E/TC:0 0 Core data-abort at address 0x6000000456c4ea0 (tag check fault)
E/TC:0 0  esr 0x96000051  ttbr0 0x0e1a9000   ttbr1 0x00000000   cidr 0x0
E/TC:0 0  cpu #0          cpsr 0x800003c4
E/TC:0 0  x0  06000000456c4ea0 x1  0000000000000001
E/TC:0 0  x2  0000000000000000 x3  0000000000000010
E/TC:0 0  x4  09000000456c4e40 x5  0000000000000010
E/TC:0 0  x6  ffffffffffffffff x7  09000000456c4e60
E/TC:0 0  x8  0000000000000020 x9  00000000456b4e90
E/TC:0 0  x10 000000004568d118 x11 0000000000000258
E/TC:0 0  x12 00000000456d906c x13 0000000000000258
E/TC:0 0  x14 00000000456d9164 x15 000000004568d118
E/TC:0 0  x16 0000000045684bd4 x17 0000000000000000
E/TC:0 0  x18 00000000456d9134 x19 0000000000000258
E/TC:0 0  x20 000000004568d118 x21 0000000000000000
E/TC:0 0  x22 0000000000000000 x23 000000004568d3d0
E/TC:0 0  x24 0000000000000000 x25 00000000456a5000
E/TC:0 0  x26 00000000456a5588 x27 00000000456a5608
E/TC:0 0  x28 00000000456a39c0 x29 00000000456d90a0
E/TC:0 0  x30 000000004563ea0c elr 000000004563ea24
E/TC:0 0  sp_el0 00000000456d90a0
E/TC:0 0 TEE load address @ 0x4562b000
E/TC:0 0 Call stack:
E/TC:0 0  0x4563ea24 add_node_to_probe at optee_os/core/kernel/dt_driver.c:559
E/TC:0 0  0x4563eb7c parse_node at optee_os/core/kernel/dt_driver.c:650
E/TC:0 0  0x45642398 call_initcalls at optee_os/core/kernel/initcall.c:41
E/TC:0 0  0x45633798 boot_init_primary_late at optee_os/core/arch/arm/kernel/boot.c:1395
E/TC:0 0 Panic '[abort] Tag check fault! (trap CPU)' at core/arch/arm/kernel/abort.c:527 <get_fault_type>
E/TC:0 0 TEE load address @ 0x4562b000
E/TC:0 0 Call stack:
E/TC:0 0  0x45633afc print_kernel_stack at optee_os/core/arch/arm/kernel/unwind_arm64.c:89
E/TC:0 0  0x456407b0 __do_panic at optee_os/core/kernel/panic.c:24
E/TC:0 0  0x456327c4 get_fault_type at optee_os/core/arch/arm/kernel/abort.c:488
E/TC:0 0  0x4562fe5c el1_sync_abort at optee_os/core/arch/arm/kernel/thread_a64.S:957
```
Tested with `optee_os` version `3.19.0-245-g7c1027426`

**Comment by jenswi-linaro** [comment]
https://github.com/OP-TEE/optee_os/pull/5760 should fix this.

**Comment by jforissier** [comment]
> #5760 should fix this.

Hmmm no, I still get the same crash :cry: 

```
I/TC: OP-TEE version: 3.19.0-228-g1d491b6c9 (gcc version 10.2.1 20201103 (GNU Toolchain for the A-profile Architecture 10.2-2020.11 (arm-10.16))) #1 Thu Jan 12 13:19:00 UTC 2023 aarch64
...
D/TC:0 0 boot_init_primary_late:1383 Memory tagging enabled
...
D/TC:0 0 call_initcalls:40 level 1 register_time_source()
D/TC:0 0 call_initcalls:40 level 1 teecore_init_pub_ram()
D/TC:0 0 call_initcalls:40 level 2 probe_dt_drivers_early()
D/TC:0 0 add_node_to_probe:553 element: dt-test-consumer on node dt-test-consumer
D/TC:0 0 add_node_to_probe:553 element: dt-test-bus-b0 on node simple-bus
D/TC:0 0 add_node_to_probe:553 element: dt-test-crypt-consumer on node dt-test-crypt-consumer
E/TC:0 0
E/TC:0 0 Core data-abort at address 0x6000000267c4ea0 (tag check fault)
E/TC:0 0  esr 0x96000051  ttbr0 0x0e1ad000   ttbr1 0x00000000   cidr 0x0
E/TC:0 0  cpu #0          cpsr 0x800003c4
E/TC:0 0  x0  06000000267c4ea0 x1  0000000000000001
E/TC:0 0  x2  0000000000000000 x3  0000000000000010
E/TC:0 0  x4  09000000267c4e70 x5  0000000000000010
E/TC:0 0  x6  ffffffffffffffff x7  09000000267c4e60
E/TC:0 0  x8  09000000267c4e60 x9  00000000267d8df0
E/TC:0 0  x10 000000002678c110 x11 0000000000000000
E/TC:0 0  x12 00000000267d905c x13 00000000267931a3
E/TC:0 0  x14 0000000000000020 x15 000000002678c110
E/TC:0 0  x16 0000000026738d8c x17 0000000000000000
E/TC:0 0  x18 00000000267d9134 x19 0000000000000258
E/TC:0 0  x20 000000002678c110 x21 00000000267a5600
E/TC:0 0  x22 09000000267c4e40 x23 0000000000000000
E/TC:0 0  x24 000000002678c3c8 x25 0000000000000000
E/TC:0 0  x26 0000000000000000 x27 00000000267a5580
E/TC:0 0  x28 00000000267a3cc8 x29 00000000267d9090
E/TC:0 0  x30 000000002673b24c elr 000000002673b264
E/TC:0 0  sp_el0 00000000267d9090
E/TC:0 0 TEE load address @ 0x26727000
E/TC:0 0 Call stack:
E/TC:0 0  0x2673b264 add_node_to_probe at optee_os/core/kernel/dt_driver.c:559
E/TC:0 0  0x2673b3d8 parse_node at optee_os/core/kernel/dt_driver.c:650
E/TC:0 0  0x2673ee1c call_initcalls at optee_os/core/kernel/initcall.c:41
E/TC:0 0  0x2672fa94 boot_init_primary_late at optee_os/core/arch/arm/kernel/boot.c:1395
E/TC:0 0 Panic '[abort] Tag check fault! (trap CPU)' at core/arch/arm/kernel/abort.c:527 <get_fault_type>
E/TC:0 0 TEE load address @ 0x26727000
E/TC:0 0 Call stack:
E/TC:0 0  0x2672fe58 print_kernel_stack at optee_os/core/arch/arm/kernel/unwind_arm64.c:89
E/TC:0 0  0x2673d058 __do_panic at optee_os/core/kernel/panic.c:24
E/TC:0 0  0x2672e8bc get_fault_type at optee_os/core/arch/arm/kernel/abort.c:488
E/TC:0 0  0x2672be98 el1_sync_abort at optee_os/core/arch/arm/kernel/thread_a64.S:957
```

**Comment by jenswi-linaro** [comment]
It seems to be a problem with QEMU 7.2, using QEMU 7.0 works.
I'll try to debug QEMU to get some clue on what's happening it seems that memory tags are corrupted in some way.

**Comment by jenswi-linaro** [comment]
Stepping through the stg instruction didn't give much. I'll try bisecting instead.

**Comment by jenswi-linaro** [comment]
I've bisected to commit b8967ddf393a ("target/arm: Use probe_access_full for MTE") in QEMU between 7.1.0 and 7.2.0. It's clearly related to MTE so this may very well be the culprit. I'll report it on the QEMU list. I guess it will take some time to get this sorted.

I'm going check that memtag works on FVP before reporting this to be sure.

**Comment by jforissier** [comment]
Thanks for investigating. In the meantime let's keep this as a known issue with v7.2.0 -- it obviously doesn't justify reverting back our manifest to v7.0.0.
Note that I intend to add CI tests with BTI + PAC + MTE in the near future, but even if QEMU isn't fixed it will not be difficult to switch to v7.0.0 in the build job.

**Comment by jenswi-linaro** [comment]
> Thanks for investigating. In the meantime let's keep this as a known issue with v7.2.0 -- it obviously doesn't justify reverting back our manifest to v7.0.0.

Agreed. I can confirm that memtag works with FVP **Base** model. It's not implemented in the Foundation model.

> Note that I intend to add CI tests with BTI + PAC + MTE in the near future, but even if QEMU isn't fixed it will not be difficult to switch to v7.0.0 in the build job.

Sounds good.

**Comment by jenswi-linaro** [comment]
Already reported by someone else at https://gitlab.com/qemu-project/qemu/-/issues/1416
